raft topology: assign tokens after join node response rpc

piodul opened the pull request:
Currently, when the topology coordinator accepts a node, it moves it to bootstrap state and assigns tokens to it (either new ones during bootstrap, or the replaced node's tokens). Only then it contacts the joining node to tell it about the decision and let it perform a read barrier.

However, this means that the tokens are inserted too early. After inserting the tokens the cluster is free to route write requests to it, but it might not have learned about all of the schema yet.

Fix the issue by inserting the tokens later, after completing the join node response RPC which forces the receiving node to perform a read barrier.

Refs: #15686
Fixes: #15738

mykaul commented:
Do we need an additional test here, or will this be covered by existing tests?

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - Build
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4213/
- Duration: 29 min
- Builder: monster

gleb-cloudius commented on db/system_keyspace.cc:
```diff
@@ -2506,15 +2506,17 @@ future<service::topology> system_keyspace::load_topology_state() {
             switch (nstate) {
+            case service::node_state::bootstrapping:
+                // We need to know how many tokens to generate (we only generate them
+                // at some point of the sequence)
```
What do you mean by "we only generate them at some point of the sequence"?

piodul replied:
To bootstrap the node, the topology coordinator has to go through several steps, states of the topology coordinator (that's what I meant by "sequence" here). Here I meant that generating tokens doesn't happen at the first step like it does now but a bit later.

I'll try to rephrase to make it clearer.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -2127,20 +2141,12 @@ class topology_coordinator {
                                .with_node(node.id)
                                .set("node_state", node_state::bootstrapping)
-                               .del("topology_request")
-                               .set("tokens", bootstrap_tokens);
```
So basically at some point of bootstrap process we will have a joining node without any tokens. Right?

piodul commented:
> Do we need an additional test here, or will this be covered by existing tests?

In order for this bug to reproduce, a write operation must happen at the right time during the bootstrap/replace procedure. This should be in general covered by SCT tests (and the issue was reported by a new SCT test) but I think we can also do it in pytests - we only do topology operations but don't write in parallel. I'll modify the `test_topology_ops.py` test to do writes asynchronously.

kbr-scylla commented:
In #15686 Aleks described two separate issues (https://github.com/scylladb/scylladb/issues/15686#issuecomment-1755368454.).

The id-to-ip issue is a duplicate of previously existing issue. This PR is solving the new issue where we write to bootstrapping node too early.

gleb-cloudius commented:
Yes. but by committing  the PR the test will still not pass, so it should not "fix" it. We can open a separate bug about early writes and close the original one as a dup of existing ip to id mapping problem.

piodul commented:
Hm, I wrote the test and, unfortunately, there is still an issue with replace (earlier, I only manually tested bootstrap). While inserting tokens later in the case of bootstrap does indeed help, I can still observe `no_such_keyspace` errors in the logs of the replacing node.

gleb-cloudius commented:
> Hm, I wrote the test and, unfortunately, there is still an issue with replace (earlier, I only manually tested bootstrap). While inserting tokens later in the case of bootstrap does indeed help, I can still observe `no_such_keyspace` errors in the logs of the replacing node.

If it is replacing with the same IP case it may be requests that were initiated before the old node was killed.

piodul commented:
> > Hm, I wrote the test and, unfortunately, there is still an issue with replace (earlier, I only manually tested bootstrap). While inserting tokens later in the case of bootstrap does indeed help, I can still observe `no_such_keyspace` errors in the logs of the replacing node.
> 
> If it is replacing with the same IP case it may be requests that were initiated before the old node was killed.

Right, but it's not reusing the IP. This is the replace config from the test:

```
replace_cfg = ReplaceConfig(replaced_id = servers[0].server_id, reuse_ip_addr = False, use_host_id = False)
```

piodul commented:
Ok, I forgot to update the part in `topology_state_load` which takes care of `replacing` nodes. It marks the new node as replacing in `token_metadata` a little too early. Will send the fix and the test shortly.

piodul commented:
v2:
- Fixed the issue with replace: now, `topology_state_load` calls `tmptr->add_replacing_endpoint` only after tokens are assigned to the replacing node
- Clarified one of the comments in `load_topology_state`
- Modified the `test_topology_ops` test case to continuously write during the test and check for replica errors in the logs

piodul commented:
I also created a new issue as suggested [here](https://github.com/scylladb/scylladb/pull/15724#issuecomment-1766104383) and changed the PR to fix the new issue and only refer to the old one.

piodul commented:
Ugh, the new test seems to be flaky... I ran it 100 times and it failed 12 times because of the `replica::stale_topology_exception` appearing in the logs. I suppose that writes shouldn't be fenced out during normal operations, so I'll investigate.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:x: - Unit Tests
#### Failed Tests (16/20839):
- [topology_experimental_raft.test_topology_ops.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4230/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_topology_ops_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_topology_ops.debug.1)
- [topology_experimental_raft.test_topology_ops.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4230/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_topology_ops_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_topology_ops.debug.3)
- [topology_experimental_raft.test_topology_ops.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4230/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_topology_ops_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_topology_ops.dev.1)
- [topology_experimental_raft.test_topology_ops.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4230/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_topology_ops_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_topology_ops.dev.2)
- [topology_experimental_raft.test_topology_ops.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4230/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_topology_ops_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_topology_ops.dev.3)
- [topology_experimental_raft.test_topology_ops.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4230/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_topology_ops_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_topology_ops.release.1)
- [topology_experimental_raft.test_topology_ops.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4230/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_topology_ops_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_topology_ops.release.2)
- [topology_experimental_raft.test_topology_ops.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4230/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_topology_ops_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_topology_ops.release.3)
- [test_topology_ops](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4230/testReport/junit/%28root%29/test_topology_ops/Tests___Unit_Tests___test_topology_ops) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_ops)
- [test_topology_ops](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4230/testReport/junit/%28root%29/test_topology_ops/Tests___Unit_Tests___test_topology_ops_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_topology_ops)
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4230/
- Duration: 6 hr 14 min
- Builder: monster

kostja commented on test/topology_experimental_raft/test_topology_ops.py:
```diff
@@ -49,3 +62,56 @@ async def test_topology_ops(request, manager: ManagerClient):
+    logger.info(f"Checking that node {server} had no failed mutations")
+    log = await manager.server_open_log(server.server_id)
+    occurrences = await log.grep("Failed to apply mutation from")
+    assert len(occurrences) == 0
```
This is going to be flaky. Any other reason for failing to apply a mutationand the test will fail.
Such issues are hard to test, I would personally be fine with no test for this bug.

kbr-scylla replied:
> Any other reason for failing to apply a mutationand the test will fail.

There should be no other reasons.

kostja replied:
It means any time, any other part of the code, fails to apply the mutation, this test will begin failing. This goes against the idea that tests should be owned. Do you want to own a test that someone else can break?

kbr-scylla replied:
Any test can be broken like that due to any random reason due to any random place in the code.

This test is very valuable, it is both a regression test and it allowed us to find out about a different existing problem (fencing during decommission).

kostja replied:
OK, I don't want to press hard on this. But please keep in mind once you add  a test you marry into it. We don't like removing tests, and going forward try to keep the tests around no matter how bad they are.
Another issue with grepping is that when the error message text changes, and you grep for absence, your test quietly stays passing but broken, and heats the universe.

piodul commented:
> #### Failed Tests (1/21010):
> * [test_lcs_table_promotion_major_compaction](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4354/testReport/junit/compaction_additional_test/TestLCSSSTablePromotion/Tests___Sanity_Tests___test_lcs_table_promotion_major_compaction) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_lcs_table_promotion_major_compaction)

Probably related to https://github.com/scylladb/scylladb/issues/15714 - a different test case but the same test class, and the failing assertion looks similar enough to suspect that.

This PR has nothing to do with compactions, so retriggering CI.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - Build
:x: - Sanity Tests
:white_check_mark: - Unit Tests
#### Failed Tests (1/21010):
- [test_lcs_table_promotion_major_compaction](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/4362/testReport/junit/compaction_additional_test/TestLCSSSTablePromotion/Tests___Sanity_Tests___test_lcs_table_promotion_major_compaction) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_lcs_table_promotion_major_compaction)
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4362/
- Duration: 2 hr 13 min
- Builder: spider3.cloudius-systems.com

piodul commented:
v3:
- Rebased
- Modified `test_topology_ops` to ignore fencing errors (created an issue for this: https://github.com/scylladb/scylladb/issues/15804) 
- Now, in `topology_state_load`, when handling a node without bootstrap tokens we don't touch `_sys_ks` or `tmptr` at all (i.e. moved the `if (rs.ring.has_value()) {` a bit higher)

piodul commented:
@gleb-cloudius @kbr-scylla review ping

kbr-scylla commented on test/topology_experimental_raft/test_topology_ops.py:
```diff
@@ -49,3 +62,57 @@ async def test_topology_ops(request, manager: ManagerClient):
+    logger.info(f"Checking that node {server} had no failed mutations")
+    log = await manager.server_open_log(server.server_id)
+    occurrences = await log.grep(expr="Failed to apply mutation from", \
+                                 filter_expr="replica::stale_topology_exception") # Disabled due to #15804
```
Please remember to refer to this place in the code from #15804 once the PR is merged (so whoever fixes #15804 will remember to update the test)

kbr-scylla commented:
For the future: instead of using `#<issue-no>`, use `scylladb/scylladb#<issue-no>` (and appropriate version for scylla-enterprise and so on).

I adjusted the cover letter during merge.
